### PR TITLE
fix(security): enforce trusted-origin policy on callback URLs and stop serializing Account.password

### PR DESCRIPTION
**Pull request by AprilNEA**
## Summary

Close the open-redirect and deep-defense gaps that currently exist on master. Implemented independently of PR #56 so master does not have to wait on the schema-native refactor.

- Introduce `AuthConfig::is_redirect_target_trusted` (complements the existing `is_origin_trusted`) and make every handler that builds a 302 Location or embeds a URL in an outgoing email go through it.
- Start respecting `advanced.disable_origin_check`, which was defined but never read.
- Stop serializing `Account.password` (bcrypt/argon2 hash). Today no handler leaks it, but the struct derives `Serialize`; this is defense in depth.

## What changed

**`crates/core/src/config.rs`** — new `is_redirect_target_trusted(target)` helper:

- relative paths (`/reset`): allowed
- protocol-relative (`//evil.com/x`): rejected
- absolute URLs: allowed only if origin matches `base_url` or a `trusted_origins` pattern
- any value: allowed when `advanced.disable_origin_check` is set

**Handler changes** — all consume the helper:

| Endpoint | Policy on untrusted callback | Rationale |
| --- | --- | --- |
| `GET /verify-email` | Silent fallback to JSON | User-initiated email click; friendlier than 400 |
| `POST /send-verification-email` | 400 | Backend-initiated; caller must fix config |
| `POST /change-email` | 400 | Same as above |
| `POST /forget-password` | Fall back to server base URL | Preserves existing behavior but now honours `trusted_origins` + `disable_origin_check` |
| `GET /reset-password/:token` | Drop callback before redirect | Prevents open redirect **and** reset-token exfiltration |
| `POST /sign-in/social`, `POST /link-social` | 400 | Stricter than upstream TS; see **breaking change** below |

**`crates/core/src/types.rs`** — `Account.password` gains `#[serde(skip_serializing, default)]`.

## Breaking change (OAuth callback)

`POST /sign-in/social` and `POST /link-social` now reject untrusted `callbackURL`. Upstream TypeScript `better-auth` does not validate at this point. If your deployment relies on a cross-subdomain OAuth callback (e.g. `auth.app.com` initiates, `accounts.app.com/cb` completes), add the callback origin to `trusted_origins` or set `advanced.disable_origin_check = true`. No existing test in this repo exercised a cross-origin path.

## Commits

- `5b6b96b` `feat(core): add AuthConfig::is_redirect_target_trusted helper`
- `4c825db` `fix(core): never serialize Account.password`
- `5f111bb` `fix(security): validate callbackURL on redirect and email-link surfaces`
- `dcc29ad` `fix(security): require trusted origin for OAuth callback_url`

## Test plan

- [x] `cargo fmt --check`
- [x] `cargo clippy --workspace -- -D warnings`
- [x] `cargo test --workspace --lib` — 293 passed
- [x] `cargo test --workspace --tests` — all green

New coverage:

- `config::tests::redirect_target_*` (7 cases: relative, protocol-relative, base_url, trusted_origins, untrusted, unparseable, disable_origin_check bypass)
- `types::tests::account_password_is_never_serialized`
- `types::tests::account_deserializes_without_password_field`
- `email_verification::tests::test_verify_email_rejects_untrusted_callback_url`
- `email_verification::tests::test_send_verification_email_rejects_untrusted_callback_url`
- `password_management::tests::test_reset_password_token_rejects_untrusted_callback_url`
- `user_management::tests::test_change_email_rejects_untrusted_callback_url`

Two existing tests (`test_verify_email_auto_sign_in_redirect_includes_cookie`, `test_verify_email_redirect_without_auto_sign_in_no_cookie`) were updated to add `myapp.com` to `trusted_origins` — previously they implicitly asserted the (insecure) "accept any callback" behavior.

## Notes

Related to PR #56 reviews (Codex P1/P2 and @AprilNEA CHANGES_REQUESTED). Implemented without any of the PR #56 abstractions (`wire::*`, `AuthSchema`, `AccountView`) so it can land independently on master.

OAuth handler coverage here is unit-test-only on the rejection path; end-to-end flow remains covered by existing integration tests.

<h3>Greptile Summary</h3>

This PR closes two security gaps on master: an open-redirect vulnerability on every handler that builds a `302 Location` or embeds a URL in an outgoing email, and defense-in-depth leakage of the `Account.password` bcrypt/argon2 hash through the `Serialize` derive.

**Key changes:**
- New `AuthConfig::is_redirect_target_trusted` and `is_absolute_trusted_callback_url` helpers centralise all callback-URL validation with layered checks: control-char/quote stripping → authority-smuggling detection (protocol-relative, backslash variants, percent-encoded `%2f`/`%5c`) → non-http/https scheme rejection (even under `disable_origin_check`) → origin comparison against `base_url` + `trusted_origins`.
- Every handler that redirects or sends email now uses one of these helpers with the correct strictness: the looser `is_redirect_target_trusted` (allows relative paths) for server-issued `302` redirects; the stricter `is_absolute_trusted_callback_url` (requires absolute http/https) for email `href` values and OAuth `redirect_uri` forwarding.
- `Account.password` gains `#[serde(skip_serializing, default)]` — the hash no longer appears in any serialized response while DB deserialization continues to work.
- `advanced.disable_origin_check` is now actually honoured (it was previously defined but never read).
- 20+ new unit tests covering all the edge cases, plus two existing redirect tests updated to reflect the now-correct same-origin policy.

<h3>Confidence Score: 5/5</h3>

All previously raised P0/P1 security concerns have been addressed in the subsequent round-2 commits; the implementation is correct, well-tested, and safe to merge.

Every concern from the prior review threads was resolved: non-http/https schemes are now blocked even under `disable_origin_check`; relative callback URLs are rejected for email/OAuth contexts via `is_absolute_trusted_callback_url`; the bare backslash authority-smuggling bypass is caught; the duplicate dead guard in `change_email_core` was dropped. The two-layer trust model (`is_redirect_target_trusted` vs `is_absolute_trusted_callback_url`) is correctly applied at every call site, the WHATWG URL parser prevents userinfo/query-string authority spoofing, IDN normalization is handled, and 293 tests pass. No new issues were found in this review pass.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| crates/core/src/config.rs | Adds `is_redirect_target_trusted`, `is_absolute_trusted_callback_url`, `extract_origin` (WHATWG URL parser), `extract_pattern_origin` (glob/IDN-aware), and `is_authority_smuggling` — the validation core for this PR; well-tested with 20+ unit cases including scheme blocks, backslash bypasses, percent-encoded bypasses, IDN normalization, and `disable_origin_check` interaction. |
| crates/core/src/types.rs | Adds `#[serde(skip_serializing, default)]` to `Account.password`; correct — `default` allows deserialization from DB rows that lack or include the column while the serialize skip prevents password hash exposure in API responses. |
| crates/api/src/plugins/email_verification/handlers.rs | POST send-verification-email validates with `is_absolute_trusted_callback_url` (email body requires absolute URL); GET verify-email uses the looser `is_redirect_target_trusted` with silent fallback — correct asymmetry since a GET is a user-clicked link where a hard 400 would be unfriendly. |
| crates/api/src/plugins/oauth/handlers.rs | `social_sign_in_core` and `link_social_core` both use `is_absolute_trusted_callback_url` — correctly stricter than the base redirect validator since OAuth providers require absolute `redirect_uri`; relative paths are rejected early before an orphaned OAuth state row is created. |
| crates/api/src/plugins/password_management/handlers.rs | `forget_password_core` uses `is_absolute_trusted_callback_url` with silent fallback (preserves email enumeration protection); `reset_password_token_core` uses the looser `is_redirect_target_trusted` with silent fallback (correct — it's a server-issued Location redirect); both handle untrusted values before any DB token creation. |
| crates/api/src/plugins/user_management/handlers.rs | `change_email_core` validates with `is_absolute_trusted_callback_url` before any DB work including the `update_without_verification` path — intentional API-contract stability even when the field isn't consumed; previous duplicate guard removed per thread reply. |
| crates/api/src/plugins/email_password.rs | `sign_up_core` and `sign_in_core` both validate `callbackURL` with `is_absolute_trusted_callback_url`; `sign_in_with_user_core` adds a redundant defence-in-depth check (well-documented comment) so callers invoked directly with a raw URL can't bypass the guard. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[callbackURL supplied] --> B{Control chars\nor double-quote?}
    B -->|yes| REJECT[Return false / 400]
    B -->|no| C{Authority\nsmuggling?\n// or backslash\nor %2f/%5c prefix}
    C -->|yes| REJECT
    C -->|no| D{disable_origin_check\n= true?}
    D -->|yes| E{Starts with /\nor extract_origin\nreturns Some?}
    E -->|no = non-http scheme| REJECT
    E -->|yes| PASS[Return true]
    D -->|no| F{Starts with /\nrelative path?}
    F -->|yes| G{is_absolute_trusted\n_callback_url mode?}
    G -->|strict - email/OAuth| REJECT
    G -->|loose - Location redirect| PASS
    F -->|no| H{extract_origin\nreturns Some?}
    H -->|no = non-http scheme| REJECT
    H -->|yes| I{is_origin_trusted?\nmatches base_url\nor trusted_origins}
    I -->|no| REJECT
    I -->|yes| PASS

    style REJECT fill:#f66,color:#fff
    style PASS fill:#6a6,color:#fff
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `crates/api/src/plugins/email_verification/handlers.rs`, line 54-61 ([link](https://github.com/better-auth-rs/better-auth-rs/blob/a98d50092c355804a6bf43d691519cb6f5d8142a/crates/api/src/plugins/email_verification/handlers.rs#L54-L61)) 

   <a href="#">P1</a> **Double question-mark when callback URL already has query parameters**

   The URL is built by unconditionally appending a question-mark separator and then the token value to the callback URL. When the caller supplies a callback URL that already contains a query string, the resulting URL ends up with two question-mark characters. Standard URL parsers treat everything after the first question-mark as the query string, so the appended token parameter will not be individually extractable — silently breaking the verification flow for those users.

   The same construction pattern appears in `forget_password_core` (password_management/handlers.rs lines 63-68), `reset_password_token_core` (lines 199 and 207), and the `create_verification_token` helper in user_management/handlers.rs (line 37).

   The fix is to check whether the callback URL already contains a question-mark before choosing the separator: use ampersand if it does, question-mark if it does not.
</details>

<sub>Reviews (6): Last reviewed commit: ["fix(security): require absolute callback..."](https://github.com/better-auth-rs/better-auth-rs/commit/9a47a9a5dbee3922b57db48792dc257e1505b0bb) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28126669)</sub>

### CODE REVIEW

File: crates/api/src/plugins/user_management/handlers.rs [hunk 1]
```diff
@@ -119,6 +132,18 @@ pub(crate) async fn change_email_core<DB: DatabaseAdapter>(
...
+    // function, but keep it adjacent to the `create_verification_token`
+    // call so future splits of this function can't silently drop the
+    // guard. The helper is idempotent on valid input.
+    if let Some(ref url) = body.callback_url
+        && !ctx.config.is_redirect_target_trusted(url)
+    {
+        return Err(AuthError::bad_request(
+            "callbackURL is not a trusted origin",
+        ));
+    }
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply crates/api/src/plugins/user_management/handlers.rs]: Keeping the second `is_redirect_target_trusted` check as a deliberate defence-in-depth guard rather than converting to `debug_assert!`. Reasons: (1) `debug_assert!` compiles out in release, so any future refactor that splits this function and accidentally drops the top-of-function check would silently ship an open-redirect regression; (2) the runtime cost is a single origin comparison, which is insignificant compared to the surrounding DB work; (3) the code comment explicitly calls out the duplication so future maintainers see it is intentional. I prefer the redundancy over a future dead-code surprise.
